### ISSUE TITLE
Open semver range does not properly unify with closed semver ranges

### ISSUE THREAD

**Issue by CAD97** [rust-lang/cargo#9029]
**Problem**

If a dependency's version is constrained to cover multiple semver-incompatible releases, the version selected does not unify with more specific restrictions from elsewhere in the build graph. E.g. a requirement for `>=0.0.0, <0.9.0` and a requirement for `^0.4.0` results in both version `0.4.x` and `0.8.y` being pulled in (or `0.4.x` and `0.1.0` with `-Zminimal-versions`), rather than unifying both versions to just `0.4.x`.

Manually adjusting the lockfile to use `0.4.x` in both cases works as expected; this is just an issue with the "find the best solution" codepath, not the "check the lockfile satisfies" codepath.

Originally reported by @MaximilianKoestler on URLO: https://users.rust-lang.org/t/crate-interoperability-and-3rd-party-types-in-interfaces/53431

**Steps**

Repro from @MaximilianKoestler : https://github.com/MaximilianKoestler/crate-version-testing

Using the `rgb` crate as our target, this is exemplified using just the three packages:

```toml
[package]
name = "lib_a"

[dependencies]
rgb = ">=0.4.0, <0.5.0"
```

```toml
[package]
name = "lib_b"

[dependencies]
rgb = ">=0.0.0, <0.9.0"
```

```toml
[package]
name = "bin_c"

[dependencies]
lib_a = { path = "../lib_a" }
lib_b = { path = "../lib_b" }
```

The generated lockfile:

```toml
# irrelevant parts removed and reordered
[[package]]
name = "lib_a"
dependencies = [
 "rgb 0.4.0",
]

[[package]]
name = "lib_b"
dependencies = [
 "rgb 0.8.25",
]

[[package]]
name = "bin_c"
dependencies = [
 "lib_a",
 "lib_b",
]

[[package]]
name = "rgb"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"

[[package]]
name = "rgb"
version = "0.8.25"
source = "registry+https://github.com/rust-lang/crates.io-index"
```

**Possible Solution(s)**

Ideally, if all constraints overlap, only a single version should be selected. Otherwise, maintaining support for old semver-incompatible versions of public dependencies (when the subset you're using didn't change) is almost meaningless, as cargo won't unify the dependency off of the most recent semver-compatible range (or first with `-Zminimal-versions`).

**Notes**

Output of `cargo version`: `cargo 1.50.0-nightly (75d5d8cff 2020-12-22)` (also occurs on stable)

**Comment by Eh2406** [comment]
Thank you for the clear description and reproducible example!
Cargos resolver (with in the one-per-major constraints) attempts to optimize for "getting everyone the most recent they can use" and does not optimize for "have the fewest versions".
I can not think of how to make the current resolver flexible over this optimization target. If pubgrub gets incorporated into Cargo things may be more flexible.

It may be possible to make an iterative optimization procedure.

**Comment by CAD97** [comment]
> Cargos resolver (with in the one-per-major constraints) attempts to optimize for "getting everyone the most recent they can use" and does not optimize for "have the fewest versions".

Understandable, and almost certainly the correct target for private dependencies. Just very unfortunate for public dependencies, where it's not so much wanting to "have the fewest versions" as wanting not to "expected type `rgb::RGBA`, found type `rgb::RGBA`".

Perhaps this is another thing that could/should be tied to [public/private dependencies](https://github.com/rust-lang/rust/issues/44663). Currently, marking the dependencies as public [panics on unimplemented `ConflictReason::PublicDependency`](https://github.com/rust-lang/cargo/blob/master/src/cargo/core/resolver/errors.rs#L182).

**Comment by Eh2406** [comment]
Yes, public/private dependencies would be a good more general solution for this problem. If you hit that panic then there is not solution for those constraints or there is a bug in resolving public/private dependencies ( which would not be surprising, there is a reason it is unstable).

**Comment by CAD97** [comment]
Definitely some bugs remaining; manually creating a lockfile with just `rgb:0.4.0` works just fine.

**Comment by wpd** [comment]
I wonder if this issue could be addressed by extending the `[patch]` feature of Cargo.
Maybe something like this? (for the example above)
```
[patch.crates.io]
lib_a.dependencies.rgb = "=0.4.0"
lib_b.dependencies.rgb = "=0.4.0"
```
This would patch the `rgb` dependency in the `lib_a` crate to say "force the version to 0.4.0".  Similar for `lib_b`'s dependency on `rgb`.  I like this approach because it is clear that the programmer is patching something that has already been published, and therefore the programmer is taking responsibility for ensuring that the `lib_a` and `lib_b` crates function properly in her/his application.

**Comment by pkgw** [comment]
Perhaps this should be a different issue, but the [Dependency Resolver][1] section of the Cargo Book doesn't mention this aspect of the current resolver, which led me to be very confused about another instance of the behavior described here. Since it seems like this behavior will continue to happen for a while, it "would be nice"™ if the book discussed it.

[1]: https://doc.rust-lang.org/cargo/reference/resolver.html

**Comment by Eh2406** [comment]
As the one who was just looking for this information, and having trouble finding it, you are in the best position to guide us on how to improve the documentation.
Where would you expect this section to go? What would you expect it to say?
If you have the time, documentation PR's are always welcome.

**Comment by pkgw** [comment]
PR submitted as #11604.

**Comment by Skgland** [comment]
I noticed that running `cargo update -p lib_b`[^0] appears to help (in case of the example[^1]), maybe cargo prefers versions already used in the dependency tree when updating individual packages. 
But running `cargo update` un-does it again.

```shell
$> cargo update -p lib_b
     Locking 0 packages to latest compatible versions
    Removing regex-syntax v0.8.6
$> cargo update
    Updating crates.io index
     Locking 1 package to latest compatible version
      Adding regex-syntax v0.8.6
```
[^0]: might be useful to others that want to remove such duplicates from the lock-file without manually editing it
[^1]: with `rgb` replaced with `regex-syntax` as old versions have been yanked and it is no longer usable as a reproducer. In `lib_a` replace the `rgb` dependency with `regex-syntax = ">=0.7.0, <0.8.0"` and in `lib_b` replace it with `regex-syntax = ">=0.0.0, <0.9.0"`.

**Comment by bergkvist** [comment]
I'm also running into this being an issue:
```toml
[dependencies]
openvm-client-executor = { git = "https://github.com/axiom-crypto/openvm-reth-benchmark.git", rev = "d159469" }
```
which fails with
```
error[E0277]: the trait bound `alloy_trie::TrieAccount: From<GenesisAccount>` is not satisfied
  --> /Users/tobias/.cargo/git/checkouts/reth-e231042ee7db3fb7/127595e/crates/chainspec/src/spec.rs:74:21
   |
74 |         state_root: state_root_ref_unhashed(&genesis.alloc),
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<GenesisAccount>` is not implemented for `alloy_trie::TrieAccount`
   |
   = note: required for `GenesisAccount` to implement `Into<alloy_trie::TrieAccount>`
note: required by a bound in `alloy_trie::root::state_root_ref_unhashed`
  --> /Users/tobias/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/alloy-trie-0.8.1/src/root.rs:92:43
   |
92 |     pub fn state_root_ref_unhashed<'a, A: Into<TrieAccount> + Clone + 'a>(
   |                                           ^^^^^^^^^^^^^^^^^ required by this bound in `state_root_ref_unhashed`

   Compiling reth-consensus v1.4.8 (https://github.com/paradigmxyz/reth?tag=v1.4.8#127595e2)
For more information about this error, try `rustc --explain E0277`.
error: could not compile `reth-chainspec` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```
essentially making it a huge hassle getting the code to compile, without manually going through the `Cargo.lock` file to unify dependencies, and never being able to run `cargo update` because it will break the code.

I believe prioritizing the latest version over unification (by default) is a mistake.
